File: packages/vscode-extension/src/auto-exec-guard.js
/**
 * Auto-exec config guard (P2 #13) — pure classification of workspace files that
 * can lead to CODE EXECUTION without an explicit "run" action, so the IDE can
 * warn once before the agent (which may trigger them via tasks / hooks / MCP /
 * run configs) starts working in a freshly-opened, untrusted workspace.
 *
 * This is the IDE-layer complement to the cc agent's own per-write permission
 * gates: the agent confirms before WRITING these files; this warns the human
 * that the OPEN workspace already CONTAINS them.
 *
 * Pure + host-free so it unit-tests without vscode. Matching is path-separator-
 * and case-insensitive (Windows-safe) and works on workspace-RELATIVE paths.
 */

// category → human label + severity (higher = louder). Kept small + explicit;
// an unknown path is simply not risky (null), never a false positive.
const CATEGORIES = {
  "mcp-config": { label: "MCP server config (can spawn processes)", severity: 5 },
  "git-hook": { label: "Git / husky hook (runs on git actions)", severity: 5 },
  "shell-profile": { label: "Shell profile (runs on new shells)", severity: 5 },
  "vscode-tasks": { label: "VS Code task (can auto-run on folder open)", severity: 4 },
  "jetbrains-run-config": { label: "JetBrains run configuration", severity: 4 },
  "vscode-launch": { label: "VS Code launch/debug config", severity: 3 },
  "vscode-settings": { label: "VS Code workspace settings", severity: 2 },
  "jetbrains-project": { label: "JetBrains project config (.idea)", severity: 2 },
};

/** Normalize to lower-case, forward-slash, no leading `./` or `/`. */
function norm(relPath) {
  return String(relPath || "")
    .replace(/\\/g, "/")
    .replace(/^\.?\/+/, "")
    .toLowerCase();
}

/**
 * Classify a workspace-relative path. Returns `{ category, label, severity }`
 * for an auto-exec-risky config file, or null for anything ordinary.
 */
function classifyAutoExecTarget(relPath) {
  const p = norm(relPath);
  if (!p) return null;
  const base = p.slice(p.lastIndexOf("/") + 1);

  let category = null;
  // MCP configs (agent auto-loads these → can spawn arbitrary servers).
  if (base === "mcp.json" || base === ".mcp.json") category = "mcp-config";
  else if (p.endsWith("/mcp.json") && (p.startsWith(".vscode/") || p.startsWith(".cursor/")))
    category = "mcp-config";
  // Git / husky hooks.
  else if (p.startsWith(".git/hooks/") || p.startsWith(".husky/")) {
    // .git/hooks/*.sample are inert templates git ships — not risky.
    if (!base.endsWith(".sample") && base !== "_" && base) category = "git-hook";
  }
  // Shell profiles (POSIX + PowerShell).
  else if (
    [
      ".bashrc",
      ".bash_profile",
      ".zshrc",
      ".zprofile",
      ".profile",
      ".kshrc",
      ".cshrc",
    ].includes(base) ||
    /microsoft\.powershell_profile\.ps1$/.test(base) ||
    base === "profile.ps1"
  )
    category = "shell-profile";
  // VS Code workspace configs.
  else if (p === ".vscode/tasks.json") category = "vscode-tasks";
  else if (p === ".vscode/launch.json") category = "vscode-launch";
  else if (p === ".vscode/settings.json") category = "vscode-settings";
  // JetBrains project configs.
  else if (p.startsWith(".idea/runconfigurations/")) category = "jetbrains-run-config";
  else if (p.startsWith(".idea/")) category = "jetbrains-project";

  if (!category) return null;
  const meta = CATEGORIES[category];
  return { category, label: meta.label, severity: meta.severity };
}

/**
 * Classify a list of workspace-relative paths into deduped findings, sorted by
 * severity (loudest first) then path. Input is whatever cheap listing the host
 * can produce (a shallow scan of the workspace root is enough for the common
 * cases; hooks live one level deep).
 */
function scanAutoExecConfig(relPaths) {
  const seen = new Set();
  const out = [];
  for (const rp of relPaths || []) {
    const hit = classifyAutoExecTarget(rp);
    if (!hit) continue;
    const key = norm(rp);
    if (seen.has(key)) continue;
    seen.add(key);
    out.push({ path: norm(rp), ...hit });
  }
  out.sort((a, b) => b.severity - a.severity || a.path.localeCompare(b.path));
  return out;
}

/** One-line human summary for a confirmation prompt (empty string when none). */
function summarizeAutoExecScan(findings) {
  if (!findings || !findings.length) return "";
  const top = findings.slice(0, 6).map((f) => `• ${f.path} — ${f.label}`);
  const more =
    findings.length > top.length ? `\n…and ${findings.length - top.length} more` : "";
  return (
    `This workspace contains ${findings.length} auto-executable config file(s). ` +
    `The agent may trigger these while it works:\n` +
    top.join("\n") +
    more
  );
}

module.exports = {
  classifyAutoExecTarget,
  scanAutoExecConfig,
  summarizeAutoExecScan,
  AUTO_EXEC_CATEGORIES: CATEGORIES,
};

File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/AutoExecGuard.java
package com.chainlesschain.ide;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

/**
 * Auto-exec config guard (P2 #13) — the JetBrains twin of the VS Code
 * extension's {@code auto-exec-guard.js}. Pure classification of workspace files
 * that can lead to CODE EXECUTION without an explicit "run" action (MCP configs,
 * git/husky hooks, shell profiles, VS Code tasks/launch, JetBrains run configs,
 * .idea project config), so the IDE can warn once before the agent — which can
 * trigger them via tasks / hooks / MCP — starts working in a freshly-opened
 * workspace. SDK-free → smoke-testable.
 *
 * <p>Matching is path-separator- and case-insensitive (Windows-safe) over
 * workspace-RELATIVE paths. An unknown path is simply not risky (null).
 */
public final class AutoExecGuard {
    private AutoExecGuard() {}

    /** A classified finding. */
    public static final class Finding {
        public final String path;
        public final String category;
        public final String label;
        public final int severity;

        Finding(String path, String category, String label, int severity) {
            this.path = path;
            this.category = category;
            this.label = label;
            this.severity = severity;
        }
    }

    private static final Map<String, String> LABELS = new LinkedHashMap<>();
    private static final Map<String, Integer> SEVERITY = new LinkedHashMap<>();

    static {
        put("mcp-config", "MCP server config (can spawn processes)", 5);
        put("git-hook", "Git / husky hook (runs on git actions)", 5);
        put("shell-profile", "Shell profile (runs on new shells)", 5);
        put("vscode-tasks", "VS Code task (can auto-run on folder open)", 4);
        put("jetbrains-run-config", "JetBrains run configuration", 4);
        put("vscode-launch", "VS Code launch/debug config", 3);
        put("vscode-settings", "VS Code workspace settings", 2);
        put("jetbrains-project", "JetBrains project config (.idea)", 2);
    }

    private static void put(String cat, String label, int sev) {
        LABELS.put(cat, label);
        SEVERITY.put(cat, sev);
    }

    private static final Set<String> SHELL_PROFILES = Set.of(
            ".bashrc", ".bash_profile", ".zshrc", ".zprofile", ".profile",
            ".kshrc", ".cshrc");

    /** Normalize to lower-case, forward-slash, no leading {@code ./} or {@code /}. */
    private static String norm(String relPath) {
        if (relPath == null) return "";
        String p = relPath.replace('\\', '/');
        p = p.replaceFirst("^\\.?/+", "");
        return p.toLowerCase(Locale.ROOT);
    }

    /** Classify a workspace-relative path, or null for anything ordinary. */
    public static Finding classify(String relPath) {
        String p = norm(relPath);
        if (p.isEmpty()) return null;
        String base = p.substring(p.lastIndexOf('/') + 1);

        String category = null;
        if (base.equals("mcp.json") || base.equals(".mcp.json")) {
            category = "mcp-config";
        } else if (p.endsWith("/mcp.json")
                && (p.startsWith(".vscode/") || p.startsWith(".cursor/"))) {
            category = "mcp-config";
        } else if (p.startsWith(".git/hooks/") || p.startsWith(".husky/")) {
            if (!base.endsWith(".sample") && !base.equals("_") && !base.isEmpty()) {
                category = "git-hook";
            }
        } else if (SHELL_PROFILES.contains(base)
                || base.endsWith("microsoft.powershell_profile.ps1")
                || base.equals("profile.ps1")) {
            category = "shell-profile";
        } else if (p.equals(".vscode/tasks.json")) {
            category = "vscode-tasks";
        } else if (p.equals(".vscode/launch.json")) {
            category = "vscode-launch";
        } else if (p.equals(".vscode/settings.json")) {
            category = "vscode-settings";
        } else if (p.startsWith(".idea/runconfigurations/")) {
            category = "jetbrains-run-config";
        } else if (p.startsWith(".idea/")) {
            category = "jetbrains-project";
        }

        if (category == null) return null;
        return new Finding(p, category, LABELS.get(category), SEVERITY.get(category));
    }

    /** Classify a list of relative paths into deduped findings, loudest-first. */
    public static List<Finding> scan(List<String> relPaths) {
        Set<String> seen = new LinkedHashSet<>();
        List<Finding> out = new ArrayList<>();
        if (relPaths != null) {
            for (String rp : relPaths) {
                Finding f = classify(rp);
                if (f == null || seen.contains(f.path)) continue;
                seen.add(f.path);
                out.add(f);
            }
        }
        out.sort((a, b) -> {
            if (a.severity != b.severity) return b.severity - a.severity;
            return a.path.compareTo(b.path);
        });
        return out;
    }

    /** One-line human summary for a confirmation prompt (empty when none). */
    public static String summarize(List<Finding> findings) {
        if (findings == null || findings.isEmpty()) return "";
        StringBuilder sb = new StringBuilder();
        sb.append("This workspace contains ").append(findings.size())
                .append(" auto-executable config file(s). ")
                .append("The agent may trigger these while it works:\n");
        int shown = Math.min(6, findings.size());
        for (int i = 0; i < shown; i++) {
            Finding f = findings.get(i);
            sb.append("• ").append(f.path).append(" — ").append(f.label);
            if (i < shown - 1) sb.append("\n");
        }
        if (findings.size() > shown) {
            sb.append("\n…and ").append(findings.size() - shown).append(" more");
        }
        return sb.toString();
    }
}
